src/table_properties.rs

```diff
@@ -0,0 +1,315 @@
+    // Other fields will be exposed as needed.
+
+    // We own the underlying pointer wrapper struct
+    table_properties: *const rocksdb_table_properties_t,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl Drop for TableProperties<'_> {
+    fn drop(&mut self) {
+        // Free the table properties holder struct
+        unsafe { ffi::rocksdb_table_properties_destroy(self.table_properties) };
+    }
+}
+
+impl<'a> TableProperties<'a> {
+    pub(crate) fn from_ptr(
+        table_properties_ptr: *const rocksdb_table_properties_t,
+    ) -> TableProperties<'a> {
+        TableProperties {
+            table_properties: table_properties_ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    // This method wraps access to a custom user-collected property via a read-only
+    // pointer to the RocksDB FFI rocksdb_table_properties_t struct.
+    pub fn get_user_collected_property(&self, key: &str) -> Option<String> {
+        unsafe {
+            let user_collected_properties =
+                ffi::rocksdb_table_properties_get_user_collected_properties(self.table_properties);
+
+            let key_cstring = CString::new(key).unwrap();
+            let value_ptr = ffi::rocksdb_user_collected_properties_get(
+                user_collected_properties,
+                key_cstring.as_ptr(),
+            );
+
+            let value = if value_ptr.is_null() {
+                None
+            } else {
+                let value_string = CStr::from_ptr(value_ptr).to_str().unwrap().to_owned();
+                ffi::rocksdb_free(value_ptr as *mut c_void);
+                Some(value_string)
+            };
+
+            ffi::rocksdb_user_collected_properties_destroy(user_collected_properties);
+
+            value
+        }
+    }
+}
+
+#[repr(C)]
+pub enum EntryType {
+    EntryPut,
+    EntryDelete,
+    EntrySingleDelete,
+    EntryMerge,
+    EntryRangeDeletion,
+    EntryBlobIndex,
+    EntryDeleteWithTimestamp,
+    EntryWideColumnEntity,
+    EntryTimedPut,
+    EntryOther,
+}
+
+#[repr(C)]
+pub struct TablePropertiesCollectorContext {
+    pub column_family_id: u32,
+    pub level_at_creation: i32,
+    pub num_levels: i32,
+    pub last_level_inclusive_max_seqno_threshold: u64,
+}
+
+/// Table properties collector factory trait
+pub trait TablePropertiesCollectorFactory {
+    type Collector: TablePropertiesCollector;
+
+    /// Create a new table properties collector
+    fn create(&mut self, context: TablePropertiesCollectorContext) -> Self::Collector;
+
+    /// Name of the collector factory to use for logging
+    fn name(&self) -> &CStr;
+}
+
+/// Table properties collector trait
+pub trait TablePropertiesCollector {
+    /// Called when a new key/value pair is added to the table
+    fn add_user_key(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        entry_type: EntryType,
+        seq: u64,
+        file_size: u64,
+    ) -> bool;
```

Comment:
Something like `Result<(), Err>` might be more appropriate here. I'll see what feels natural as I build the LSN tracking.

Reply:
How does RocksDB use the returned `Status`?